### PR TITLE
[DistillationTrainer refactor] Move IW-OPD to a dedicated experimental trainer

### PR DESCRIPTION
**Pull request by qgallouedec**
*stacked on PR #6452. Part of #6449.*

IW-OPD (Importance-Weighted On-Policy Distillation, [2606.22600](https://huggingface.co/papers/2606.22600)) is a sampled-token objective with per-token importance weights: a teacher-guided **policy-gradient** method (the paper optimizes it with a clipped PG / verl setup and vLLM rollouts), not a supervised full-vocab divergence. It fits neither the stable, full-vocabulary `DistillationTrainer` this series is converging on nor GRPO (which has no teacher). Rather than delete a working method, this PR **moves** it, verbatim, to its own self-contained home.

New trainer — `trl/experimental/iw_opd/`:

- `IWOPDTrainer` / `IWOPDConfig`, forked from the pre-refactor `DistillationTrainer` / `DistillationConfig` with `distillation_objective="iw_opd"` as its reason for existing. It is a **frozen snapshot**, flagged as such in its docstring: not actively maintained, and it will not track the refactored trainer's later improvements.
- Registered in the telemetry allowlist (`base_trainer.py`).
- Its own test module `tests/experimental/test_iw_opd_trainer.py`, forked from the distillation tests and retargeted at `IWOPD*`.

Removed from the base trainer (same deletions as a plain removal, since the logic now lives in `iw_opd/`):

- `_compute_iw_opd_loss` and both of its `compute_loss` dispatch branches (local and server);
- the `rollout_logprobs` plumbing that existed only to feed its importance ratio — the sampled-logprob extraction in `_store_completions_in_buffer`, the buffered `rollout_logprobs` tensor, and the `logprobs=0` request to the vLLM backend (now always `None`);
- config fields `distillation_objective`, `iw_opd_gamma`, `iw_opd_epsilon` and their validations;
- the corresponding tests.

`compute_loss` goes from 7 dispatch paths to 5. The generalized-JSD objective is unchanged and stays pinned by the tests added in #6450.

The `paper_index.md` entry for IW-OPD is repointed from `experimental.distillation.DistillationTrainer` to `experimental.iw_opd.IWOPDTrainer` (the GKD entry and the general OPD-blog entry stay on `DistillationTrainer`).

### Reordering / merge note

Plan step 1.3 (remove IW-OPD from the trainer) and step 1.4 (drop its paper-index entry) are merged here, and the pair is reordered to run first in Phase 1: IW-OPD was entangled *inside* the server dispatch branch (`if distillation_objective == "iw_opd"` nested under `if use_teacher_server`), so re-homing it first makes the later server extraction a smaller, cleaner move. The former plan step 10.1 ("re-home IW-OPD on a GRPO-shaped trainer") is subsumed — it is re-homed here instead, on a frozen snapshot rather than GRPO, because it is not a GRPO-shaped method.

### Risk

Medium. Deletes a working objective and its rollout-logprob plumbing from the base (which also touches the vLLM generation path, `logprobs=None`), but the logic is preserved verbatim in `iw_opd/` and covered by the forked test module (43 tests green). The change is confined to IW-OPD branches and the generated-completion buffer; the JSD paths are untouched.

---

> [!NOTE]
> **Medium Risk**
> Removes a public objective from `DistillationTrainer` (breaking for `distillation_objective="iw_opd"`), but behavior is preserved in the new module; vLLM generation changes are limited to dropping sampled logprobs on the base trainer.
> 
> **Overview**
> **Moves Importance-Weighted On-Policy Distillation (IW-OPD) out of `DistillationTrainer`** into a dedicated `trl.experimental.iw_opd` package with `IWOPDTrainer` and `IWOPDConfig` (forked snapshot, default objective `iw_opd`). `DistillationTrainer` is narrowed to generalized JSD/KL only: it drops `distillation_objective`, `iw_opd_*` config, `_compute_iw_opd_loss`, IW-OPD branches in `compute_loss`, and the vLLM rollout logprob path (`rollout_logprobs` buffering and `logprobs=0` on generate).
> 
> IW-OPD tests move from `test_distillation_trainer.py` to new `test_iw_opd_trainer.py`; docs and `paper_index.md` point IW-OPD users at `IWOPDTrainer`. `IWOPDTrainer` is added to the telemetry allowlist in `base_trainer.py`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 28d90ba3f3a97cea9ff95316442acd4d06275edd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/iw_opd/iw_opd_trainer.py [hunk 1]
```diff
@@ -0,0 +1,1911 @@
...
+    - Local teacher model or external teacher via vLLM server
+    - Student on-policy generation via vLLM or model.generate()
+    - Liger kernel for memory-efficient fused JSD loss
+    """
+
+    _tag_names = ["trl", "iw-opd"]
+    _name = "IW-OPD"
+    _paper = {
+        "title": "On-Policy Distillation of Language Models: Learning from Self-Generated Mistakes",
+        "id": "2306.13649",
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/iw_opd/iw_opd_trainer.py]: https://github.com/huggingface/trl/pull/6453/commits/456df918c68331e37fb0b0de9fc6365c933d946f

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -590,9 +590,6 @@ def __init__(
         # ── Store config values ──
         self.lmbda = args.lmbda
         self.beta = args.beta
-        self.distillation_objective = args.distillation_objective
-        self.iw_opd_gamma = args.iw_opd_gamma
-        self.iw_opd_epsilon = args.iw_opd_epsilon
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py:595]: We should also remove these from the docs in distillation_trainer.md.

File: tests/experimental/test_distillation_trainer.py [hunk 3]
```diff
@@ -800,164 +754,6 @@ def test_sampled_mode_matches_between_local_and_external_teachers(self, monkeypa
...
-            top_logprobs=1,
-        )
-
-        loss = trainer.compute_loss(trainer.model, inputs)
-
-        assert torch.isfinite(loss)
-        assert teacher_client.calls[0]["top_logprobs"] == 1
-
-
-class TestDistillationTrainerServerPath(TrlTestCase):
```
Comment (cmpatino) [inline tests/experimental/test_distillation_trainer.py:970]: Was removing the `class TestDistillationTrainerServerPath(TrlTestCase)` definition intentional?

File: trl/experimental/iw_opd/iw_opd_config.py [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+    )
+    beta: float = field(
+        default=1.0,
+        metadata={
+            "help": "Interpolation coefficient for the Generalized JSD loss. "
+            "0.0 = forward KL, 0.5 = JSD, 1.0 = reverse KL."
+        },
+    )
+    distillation_objective: str = field(
+        default="jsd",
```
Comment (cmpatino) [inline trl/experimental/iw_opd/iw_opd_config.py]: I think the default here should be `iw_opd` since we're in the IW-OPD trainer.

**Review by cmpatino — COMMENTED** [review]
Comments about args in the new trainer and a potential accidental deletion.

File: tests/experimental/test_iw_opd_trainer.py [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+        teacher_student = IWOPDTrainer.generalized_jsd_loss(self.teacher_logits, self.student_logits, beta=0.5)
+        assert student_teacher == teacher_student
+
+    def test_zero_loss_for_identical_inputs(self):
+        identical_logits = torch.randn(self.batch_size, self.seq_length, self.vocab_size)
+        loss = IWOPDTrainer.generalized_jsd_loss(identical_logits, identical_logits)
+        assert round(abs(loss.item() - 0), 6) == 0
+
+
+class TestDistillationTrainer(TrlTestCase):
```
Comment (cmpatino) [inline tests/experimental/test_iw_opd_trainer.py]: Nit: Should we change this for `TestIWOPDTrainer` to avoid confusion with the `DistillationTrainer` tests?

File: trl/experimental/distillation/distillation_trainer.py (hunk 2 above)
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:595]: Done, removed the distillation_objective bullet and the whole IW-OPD subsection from distillation_trainer.md in this PR

File: tests/experimental/test_distillation_trainer.py (hunk 3 above)
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py:970]: No, accidental clip while deleting the adjacent IW-OPD tests; it's a server-path class, unrelated to IW-OPD. Restored the class header, so setup_class/_run_one_step/test_reverse_kl_finite_grad_with_ragged_batch are their own class again. Thanks

File: trl/experimental/iw_opd/iw_opd_config.py (hunk 4 above)
Comment (qgallouedec) [reply trl/experimental/iw_opd/iw_opd_config.py]: Good point, changed it to "iw_opd".

File: tests/experimental/test_iw_opd_trainer.py (hunk 5 above)
Comment (qgallouedec) [reply tests/experimental/test_iw_opd_trainer.py]: right, renamed to TestIWOPDTrainer

File: trl/experimental/iw_opd/iw_opd_config.py [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+        metadata={
+            "help": "Objective to optimize. Use 'jsd' for the existing generalized JSD/KL objective or 'iw_opd' "
+            "for sampled-token Importance-Weighted On-Policy Distillation."
+        },
+    )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/iw_opd/iw_opd_config.py:218]: Fixed, distillation_objective now defaults to "iw_opd"; IWOPDConfig() runs the importance-weighted objective, not JSD.

File: trl/experimental/iw_opd/iw_opd_config.py [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+            "0.0 = forward KL, 0.5 = JSD, 1.0 = reverse KL."
+        },
+    )
+    distillation_objective: str = field(
+        default="iw_opd",
+        metadata={
+            "help": "Objective to optimize. Use 'iw_opd' for sampled-token Importance-Weighted On-Policy "
+            "Distillation or 'jsd' for the generalized JSD/KL objective."
+        },
+    )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/iw_opd/iw_opd_config.py:218]: Good catch, fixed by pinning distillation_objective="jsd" in the JSD-path config
